File: ruby-wallaroo/lib/wallaroo/feature.rb
# Copyright (c) 2012 Red Hat, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'wallaroo/client'

module Wallaroo
  module Client
    class Feature
      include ::Wallaroo::Client::Proxying      
      include ::Wallaroo::Client::ArcUtils
      
      [:name, :includes, :conflicts, :depends, :parameters].each do |what|
        # XXX: distinguish sensibly between readonly and read-write attributes
        declare_attribute what
      end
      
      alias included_features includes
      alias included_features= includes=

      [:name].each do |what|
        define_method "set#{Util.camelcase(what.to_s)}" do |val|
          self.attr_vals[what] = val
          self.update!
          self.refresh
        end
      end

      [[:included_features, "includes", true], 
       [:depends, "depends on", false], 
       [:conflicts, "conflicts with", false]].each do |what, explain, order_preserving|
        define_method "modify#{Util.camelcase(what.to_s)}" do |command, fset, *options|
          options = options[0] || {}
          modify_arcs(command,fset,options,what,"#{what}=",:explain=>explain, :preserve_order=>order_preserving)
          update!
        end
      end

      def modifyParams(command, params, options=nil)
        options ||= {}
        case command.upcase
        when "ADD" then
          self.parameters = self.parameters.merge(params)
        when "REMOVE" then
          self.parameters = self.parameters.reject {|k,v| params.include?(k) }
        when "REPLACE" then
          self.parameters = params
        else 
          fail(Errors.make(Errors::BAD_COMMAND, errwhat), "Invalid command #{command}")
        end
        update!
      end
    end
  end
end
      

File: python-wallaroo/wallaroo/client/feature.py
# Copyright (c) 2013 Red Hat, Inc.
# Author:  William Benton (willb@redhat.com)

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from proxy import Proxy, proxied_attr
from proxy import proxied_attr_get as pag, proxied_attr_set as pas, proxied_attr_getset as pags

from arc_utils import arcmethod, uniq

import errors
from errors import not_implemented, fail

class feature(Proxy):
    name = property(*pags("name"))
    
    included_features = property(*pags("includes"))
    includes = property(*pags("includes")) #convenience alias
    
    depends = property(*pags("depends"))
    conflicts = property(*pags("conflicts"))
    parameters = property(*pags("parameters"))
    
    modifyIncludes = arcmethod(*pags("includes"), explain="includes", preserve_order=True)
    modifyDepends = arcmethod(*pags("depends"), explain="depends on", preserve_order=False)
    modifyConflicts = arcmethod(*pags("conflicts"), explain="conflicts with", preserve_order=False)
    
    def setName(self, nn):
        self.name = nn
        self.update()
    
    def modifyParams(command, params, **options):
        command = command.upper()
        if command == "ADD":
            for k, v in params.iteritems():
                self.parameters[k] = v
        elif command == "REMOVE":
            for k in [k for k in params if k in self.parameters]:
                del self.parameters[k]
        elif command == "REPLACE":
            self.parameters = params
        else:
            fail(errors.make(errors.BAD_COMMAND, errors.FEATURE), "Invalid command %s" % command)
        self.update()

proxied_attr(feature, "name")
proxied_attr(feature, "includes")
proxied_attr(feature, "conflicts")
proxied_attr(feature, "depends")
proxied_attr(feature, "parameters")